src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs

```diff
@@ -118,27 +121,49 @@ protected AuthenticationHandler(IOptionsMonitor<TOptions> options, ILoggerFactor
         protected string BuildRedirectUri(string targetPath)
             => Request.Scheme + "://" + Request.Host + OriginalPathBase + targetPath;
 
+        protected virtual string ResolveTarget(string scheme)
+            => scheme ?? Options.ForwardDefaultSelector?.Invoke(Context) ?? Options.ForwardDefault;
+
         public async Task<AuthenticateResult> AuthenticateAsync()
         {
-            // Calling Authenticate more than once should always return the original value.
-            var result = await HandleAuthenticateOnceAsync();
-            if (result?.Failure == null)
+            if (_inAuthenticate)
+            {
+                throw new InvalidOperationException("Authenticate for scheme:[" + Scheme.Name + "] resulted in a recursive call back to itself.");
+            }
+
+            _inAuthenticate = true;
+            try
             {
-                var ticket = result?.Ticket;
-                if (ticket?.Principal != null)
+                var target = ResolveTarget(Options.ForwardAuthenticate);
+                if (target != null && !string.Equals(target, Scheme.Name, StringComparison.Ordinal))
```

Comment:
helper method?
Why didn't you do this self-reference check in CookieAuth?

Reply:
yeah I guess I can push the name comparison into Resolve and have it return null when its self referencing